src/util/address.rs

```diff
@@ -176,49 +382,207 @@ impl FromStr for AddressType {
+    /// [`opcodes::all::OP_PUSHNUM_7`])
+    V7 = 7,
+
+    /// Version 8 of witness program, represented by 0x58 byte (corresponds to
+    /// [`opcodes::all::OP_PUSHNUM_8`])
+    V8 = 8,
+
+    /// Version 9 of witness program, represented by 0x59 byte (corresponds to
+    /// [`opcodes::all::OP_PUSHNUM_9`])
+    V9 = 9,
+
+    /// Version 10 of witness program, represented by 0x5a byte (corresponds to
+    /// [`opcodes::all::OP_PUSHNUM_10`])
+    V10 = 10,
+
+    /// Version 11of witness program, represented by 0x5b byte (corresponds to
+    /// [`opcodes::all::OP_PUSHNUM_11`])
+    V11 = 11,
+
+    /// Version 12 of witness program, represented by 0x5c byte (corresponds to
+    /// [`opcodes::all::OP_PUSHNUM_12`])
+    V12 = 12,
+
+    /// Version 13 of witness program, represented by 0x5d byte (corresponds to
+    /// [`opcodes::all::OP_PUSHNUM_13`])
+    V13 = 13,
+
+    /// Version 14 of witness program, represented by 0x5e byte (corresponds to
+    /// [`opcodes::all::OP_PUSHNUM_14`])
+    V14 = 14,
+
+    /// Version 15 of witness program, represented by 0x5f byte (corresponds to
+    /// [`opcodes::all::OP_PUSHNUM_15`])
+    V15 = 15,
+
+    /// Version 16 of witness program, represented by 0x60 byte (corresponds to
+    /// [`opcodes::all::OP_PUSHNUM_16`])
+    V16 = 16,
+}
+
+impl From<FutureWitnessVersion> for WitnessVersion {
+    fn from(version: FutureWitnessVersion) -> Self {
+        match version {
+            FutureWitnessVersion::V2 => WitnessVersion::V2,
+            FutureWitnessVersion::V3 => WitnessVersion::V3,
+            FutureWitnessVersion::V4 => WitnessVersion::V4,
+            FutureWitnessVersion::V5 => WitnessVersion::V5,
+            FutureWitnessVersion::V6 => WitnessVersion::V6,
+            FutureWitnessVersion::V7 => WitnessVersion::V7,
+            FutureWitnessVersion::V8 => WitnessVersion::V8,
+            FutureWitnessVersion::V9 => WitnessVersion::V9,
+            FutureWitnessVersion::V10 => WitnessVersion::V10,
+            FutureWitnessVersion::V11 => WitnessVersion::V11,
+            FutureWitnessVersion::V12 => WitnessVersion::V12,
+            FutureWitnessVersion::V13 => WitnessVersion::V13,
+            FutureWitnessVersion::V14 => WitnessVersion::V14,
+            FutureWitnessVersion::V15 => WitnessVersion::V15,
+            FutureWitnessVersion::V16 => WitnessVersion::V16,
+        }
+    }
+}
+
+impl FutureWitnessVersion {
+    // Private method used only internally for WitnessVersion -> FutureWitnessVersion conversions
+    // If anyone will try to make this method public, please remove the panic and replace the return
+    // type with `Result` - or just implement `TryFrom` trait instead, if MSRV will allow.
+    fn with(version: &WitnessVersion) -> Self {
+        match version {
+            WitnessVersion::V0 |
+            WitnessVersion::V1 => unreachable!("internal constructor `FutureWitnessVersion::with` was called on future witness versions"),
```

Comment:
`dont_panic`/`no_panic` candidate?

Reply:
Not sure I understand. If you are talking about https://docs.rs/no-panic/0.1.15/no_panic/index.html than we can't add dependencies just because of this place

Reply:
Yeah, that one. I was thinking about treating it like dev-dependency (although it can't go to `[dev-dependencies]`) and start using it at some places to prove correctness. This could be one of them.

Reply:
We need @apoelstra approval on having a new dependency

Reply:
This is kinda neat but I don't think it's worth pulling in an extra dep, especially one that does such deep magic.

Reply:
Not actually that much magic, it simply puts call to non-existing function into panicking branches, name of the function being the error message.

I wonder what kind of harm do you see from having it as optional and using it in CI only?

Reply:
With the removal of the `FutureWitnessVersion` this code is gone